docs/man/zvault-config.1.md
zvault-config(1) -- Display or change the configuration
=======================================================

## SYNOPSIS

`zvault config <REPO>`


## DESCRIPTION

This subcommand displays or changes the configuration of the repository `REPO`.
The configuration can be changes using the options described below. If no
options are set, the current configuration is displayed. Otherwise, the
configuration is changed as specified and then displayed.

Beware that the *chunker algorithm*, *chunk size* and *hash method* should not
be changed on existing repositories already containing many backups. If those
values are changed, new backups will not be able to use existing data for
deduplication. This can waste lots of storage space and most likely outweighs
the expected benefits.

The values for *bundle size*, *compression* and *encryption* only affect new
data and can be changed at any time without any drawback.


## OPTIONS

* `--bundle-size <SIZE>`:

  Set the target bundle size in MiB (default: 25).
  Please see _zvault(1)_ for more information on *bundle size*.


* `--chunker <CHUNKER>`:

  Set the chunker algorithm and target chunk size (default: fastcdc/16).
  Please see _zvault(1)_ for more information on *chunkers* and possible
  values.


* `-c`, `--compression <COMPRESSION>`:

  Set the compression method and level (default: brotli/3).
  Please see _zvault(1)_ for more information on *compression* and possible
  values.


* `-e`, `--encryption <PUBLIC_KEY>`:

  Use the given public key for encryption. The key must be a valid public key
  encoded as hexadecimal. Please use _zvault-genkey(1)_ to generate keys and
  _zvault-addkey(1)_ to add keys to the repository.

  If `none` is given as public key, encryption is deactivated.

  **Warning:** ZVault does not verify that the matching secret key which is
  needed for decryption is known.

  Please see _zvault(1)_ for more information on *encryption*.


* `--hash <HASH>`:

  Set the hash method (default: blake2).
  Please see _zvault(1)_ for more information on *hash methods* and possible
  values.


* `-q`, `--quiet`:

  Print less information


* `-v`, `--verbose`:

  Print more information


* `-h`, `--help`:

  Prints help information


* `-V`, `--version`:     

  Prints version information


## COPYRIGHT

Copyright (C) 2017-2018  Dennis Schwerdel
This software is licensed under GPL-3 or newer (see LICENSE.md)


CHANGELOG.md
# Changelog

This project follows [semantic versioning](http://semver.org).


### UNRELEASED
* [added] Translation infrastructure (**requires nightly rust**)
* [added] Checking hashes of chunks in check --bundle-data
* [added] Debian packet for libsodium23
* [modified] Updated dependencies
* [modified] Updated copyright date
* [modified] Moved all code into one crate for easier translation
* [modified] Compression ratio is now displayed in a clearer format
* [fixed] Also including the first min_size bytes in hash
* [fixed] Fixed some texts in manpages
* [fixed] Calling strip on final binaries
* [fixed] Fixed bug that caused repairs to miss some errors


### v0.4.0 (2017-07-21)
* [added] Added `copy` subcommand
* [added] Added support for xattrs in fuse mount
* [added] Added support for block/char devices
* [added] Added support for fifo files
* [modified] Reformatted sources using rustfmt
* [modified] Also documenting common flags in subcommands
* [modified] Using repository aliases (**conversion needed**)
* [modified] Remote path must be absolute
* [modified] Not removing prefixes from bundle names anymore
* [fixed] Fixed tarfile import


### v0.3.2 (2017-05-11)
* [modified] Changed order of arguments in `addkey` to match src-dst scheme
* [modified] Skip root folder on restore
* [fixed] Fixed `addkey` subcommand
* [fixed] Fixed reading tar files from stdin
* [fixed] Fixed exporting files with long names as tar files


### v0.3.1 (2017-05-09)
* [added] Derive key pairs from passwords
* [modified] Added root repository to exclude list
* [modified] Initializing data in index before use
* [modified] Updated dependencies


### v0.3.0 (2017-04-27)
* [added] Ability to read/write tar file from/to stdin/stdout
* [added] Added date to bundles
* [added] Option to combine small bundles
* [added] Fixed chunker
* [modified] Logging to stderr
* [modified] Enforce deterministic bundle ordering
* [modified] More info in analyze subcommand
* [modified] Estimating final bundle size in order to reach it
* [fixed] Only print "repairing bundles" if actually repairing bundles
* [fixed] Only put mode bits of st_mode into metadata
* [fixed] Only repairing backups with --repair
* [fixed] Fixed vacuum
* [fixed] First removing bundles, then adding new ones
* [fixed] No longer clobbering broken files


### v0.2.0 (2017-04-14)
* [added] Added CHANGELOG
* [added] Locking local repository to avoid index corruption
* [added] Storing user/group names in backups
* [added] Ability to repair bundles, backups, index, bundle map and bundle cache
* [added] Manpages for all subcommands
* [added] Folders of backups can be listed, removed and mounted
* [added] Supporting extended attributes in tar files
* [modified] No longer trying to upload by rename
* [modified] No longer failing restore if setting file attributes fails
* [modified] Backup files must end with `.backup` (**conversion needed**)
* [modified] Bundle files must end with `.bundle`
* [modified] Ignoring corrupt bundles instead of failing
* [fixed] Creating empty bundle cache on init to avoid warnings
* [fixed] Calling sodiumoxide::init for faster algorithms and thread safety (not needed)
* [fixed] Fixed a deadlock in the bundle upload code
* [fixed] Also setting repository dirty on crash
* [fixed] Ignoring missing backups folder
* [fixed] Fixed problems with uploads from relative repository paths
* [fixed] Fixed finished messages
* [fixed] Fixed inode retrieval for single-file backups
* [fixed] Fixed tar import


### v0.1.0 (2017-04-11)
First working alpha release

This release features the main functionality:
* Initializing repository
  - Generating a key on the fly
  - Import existing repository
* Creating backups
  - Partial backups
  - Deduplication
  - Compression
  - Encryption
  - From tar files
  - Support for file permissions, file date and extended attributes
* Restoring backups
  - Full or subtrees
  - To tar files
* Mounting backups or the whole repository
* Removing backups
  - Full or only specific subtrees
  - By date (`prune` subcommand)
* Check integrity
  - Repository
  - Bundles
  - Index
  - Backups
  - Inode trees
* Vacuum
  - By ratio
* Listing & Info methods
  - Repository info
  - Backup info/list
  - Directory list, Inode info
  - Bundle list and info
* Utility commands
  - `analyze`: analyze chunk usage
  - Key management commands (`addkey`, `genkey`)
  - `algotest`: algorithm testing
  - `versions`: find versions of a file
  - `diff`: Find differences between backups
  - `config`: Getting and setting config options
* Command line client
  - Powerful argument parsing
  - Nice colorful error messages
  - Progress bars
  - Man pages for main commands
* Special functionality
  - Shared repositories


src/repository/config.rs
use prelude::*;

use serde_yaml;

use std::fs::File;
use std::path::Path;
use std::io;


quick_error!{
    #[derive(Debug)]
    pub enum ConfigError {
        Io(err: io::Error) {
            from()
            cause(err)
        }
        Parse(reason: &'static str) {
            from()
            description(tr!("Failed to parse config"))
            display("{}", tr_format!("Failed to parse config: {}", reason))
        }
        Yaml(err: serde_yaml::Error) {
            from()
            cause(err)
            description(tr!("Yaml format error"))
            display("{}", tr_format!("Yaml format error: {}", err))
        }
    }
}


impl HashMethod {
    fn from_yaml(yaml: &str) -> Result<Self, ConfigError> {
        HashMethod::from(yaml).map_err(ConfigError::Parse)
    }

    fn to_yaml(&self) -> String {
        self.name().to_string()
    }
}


struct ChunkerYaml {
    method: String,
    avg_size: usize,
    seed: u64
}
impl Default for ChunkerYaml {
    fn default() -> Self {
        ChunkerYaml {
            method: "fastcdc".to_string(),
            avg_size: 16 * 1024,
            seed: 0
        }
    }
}
serde_impl!(ChunkerYaml(String) {
    method: String => "method",
    avg_size: usize => "avg_size",
    seed: u64 => "seed"
});

impl ChunkerType {
    fn from_yaml(yaml: &ChunkerYaml) -> Result<Self, ConfigError> {
        ChunkerType::from(&yaml.method, yaml.avg_size, yaml.seed).map_err(ConfigError::Parse)
    }

    fn to_yaml(&self) -> ChunkerYaml {
        ChunkerYaml {
            method: self.name().to_string(),
            avg_size: self.avg_size(),
            seed: self.seed()
        }
    }
}



impl Compression {
    #[inline]
    fn from_yaml(yaml: &str) -> Result<Self, ConfigError> {
        Compression::from_string(yaml).map_err(|_| ConfigError::Parse(tr!("Invalid codec")))
    }

    #[inline]
    fn to_yaml(&self) -> String {
        self.to_string()
    }
}


impl EncryptionMethod {
    #[inline]
    fn from_yaml(yaml: &str) -> Result<Self, ConfigError> {
        EncryptionMethod::from_string(yaml).map_err(|_| ConfigError::Parse(tr!("Invalid codec")))
    }

    #[inline]
    fn to_yaml(&self) -> String {
        self.to_string()
    }
}


struct EncryptionYaml {
    method: String,
    key: String
}
impl Default for EncryptionYaml {
    fn default() -> Self {
        EncryptionYaml {
            method: "sodium".to_string(),
            key: "".to_string()
        }
    }
}
serde_impl!(EncryptionYaml(String) {
    method: String => "method",
    key: String => "key"
});



struct ConfigYaml {
    compression: Option<String>,
    encryption: Option<EncryptionYaml>,
    bundle_size: usize,
    chunker: ChunkerYaml,
    hash: String
}
impl Default for ConfigYaml {
    fn default() -> Self {
        ConfigYaml {
            compression: Some("brotli/5".to_string()),
            encryption: None,
            bundle_size: 25 * 1024 * 1024,
            chunker: ChunkerYaml::default(),
            hash: "blake2".to_string()
        }
    }
}
serde_impl!(ConfigYaml(String) {
    compression: Option<String> => "compression",
    encryption: Option<EncryptionYaml> => "encryption",
    bundle_size: usize => "bundle_size",
    chunker: ChunkerYaml => "chunker",
    hash: String => "hash"
});



#[derive(Debug, Clone, Eq, PartialEq)]
pub struct Config {
    pub compression: Option<Compression>,
    pub encryption: Option<Encryption>,
    pub bundle_size: usize,
    pub chunker: ChunkerType,
    pub hash: HashMethod
}
impl Default for Config {
    fn default() -> Self {
        Config {
            compression: Some(Compression::from_string("brotli/3").unwrap()),
            encryption: None,
            bundle_size: 25 * 1024 * 1024,
            chunker: ChunkerType::from_string("fastcdc/16").unwrap(),
            hash: HashMethod::Blake2
        }
    }
}
serde_impl!(Config(u64) {
    compression: Option<Compression> => 0,
    encryption: Option<Encryption> => 1,
    bundle_size: usize => 2,
    chunker: ChunkerType => 3,
    hash: HashMethod => 4
});

impl Config {
    fn from_yaml(yaml: ConfigYaml) -> Result<Self, ConfigError> {
        let compression = if let Some(c) = yaml.compression {
            Some(try!(Compression::from_yaml(&c)))
        } else {
            None
        };
        let encryption = if let Some(e) = yaml.encryption {
            let method = try!(EncryptionMethod::from_yaml(&e.method));
            let key = try!(parse_hex(&e.key).map_err(|_| {
                ConfigError::Parse(tr!("Invalid public key"))
            }));
            Some((method, key.into()))
        } else {
            None
        };
        Ok(Config {
            compression,
            encryption,
            bundle_size: yaml.bundle_size,
            chunker: try!(ChunkerType::from_yaml(&yaml.chunker)),
            hash: try!(HashMethod::from_yaml(&yaml.hash))
        })
    }

    fn to_yaml(&self) -> ConfigYaml {
        ConfigYaml {
            compression: self.compression.as_ref().map(|c| c.to_yaml()),
            encryption: self.encryption.as_ref().map(|e| {
                EncryptionYaml {
                    method: e.0.to_yaml(),
                    key: to_hex(&e.1[..])
                }
            }),
            bundle_size: self.bundle_size,
            chunker: self.chunker.to_yaml(),
            hash: self.hash.to_yaml()
        }
    }

    pub fn load<P: AsRef<Path>>(path: P) -> Result<Self, ConfigError> {
        let f = try!(File::open(path));
        let config = try!(serde_yaml::from_reader(f));
        Config::from_yaml(config)
    }

    pub fn save<P: AsRef<Path>>(&self, path: P) -> Result<(), ConfigError> {
        let mut f = try!(File::create(path));
        try!(serde_yaml::to_writer(&mut f, &self.to_yaml()));
        Ok(())
    }
}
